File: problems/problem_34.py
# Find the sum of all numbers which are equal to the sum of the factorial of their digits.
def main():
    digit_factorials = []
    for number in range(3, 250000):
        if sum([factorial(digit) for digit in digits(number)]) == number:
            digit_factorials.append(number)
    print(sum(digit_factorials))

def digits(number):
    return [int(digit) for digit in str(number)]

def factorial(number):
    if number == 0:
        return 1

    factorial = 1
    for i in range(1, number + 1):
        factorial *= i
    return factorial

if __name__ == "__main__":
    main()

File: problems/problem_34.rb
# Find the sum of all numbers which are equal to the sum of the factorial of their digits.
class DigitFactorials

  private

  def initialize
    @digit_factorials = digit_factorials
  end

  def digit_factorials
    (3..250_000).each_with_object([]) do |number, list|
      list.push(number) if number.digits.map { |digit| factorial(digit) }.compact.sum == number
    end
  end

  def factorial(number)
    return 1 if number == 0

    (1..number).inject(:*)
  end

  public

  def sum
    @digit_factorials.sum
  end

end

